text/3536-unsized-thin-pointers.md

```diff
@@ -0,0 +1,286 @@
+- Feature Name: `unsized_thin_pointers`
+- Start Date: 2023-11-29
+- RFC PR: [rust-lang/rfcs#3536](https://github.com/rust-lang/rfcs/pull/3536)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Enable user code to define dynamically-sized thin pointers. Such types are
+`!Sized`, but references to them are pointer-sized (i.e. not "fat pointers").
+The implementation of [`core::mem::size_of_val()`](size_of_val) delegates to
+a new `core::mem::DynSized` trait at runtime.
+
+[size_of_val]: https://doc.rust-lang.org/core/mem/fn.size_of_val.html
+
+# Motivation
+[motivation]: #motivation
+
+Enable ergonomic and efficient references to dynamically-sized values that
+are capable of computing their own size.
+
+It should be possible to declare a Rust type that is `!Sized`, but has
+references that are pointer-sized and therefore only require a single register
+on most architectures.
+
+In particular this RFC aims to support a common pattern in other low-level
+languages, such as C, where a value may consist of a fixed-layout header
+followed by dynamically-sized data:
+
+```c
+struct __attribute__((aligned(8))) request {
+    uint32_t size;
+    uint16_t id; 
+    uint16_t flags;
+    /* uint8_t request_data[]; */
+};
+
+void handle_request(struct request *req) { /* ... */ }
+```
+
+This pattern is used frequently in zero-copy APIs that transmit structured data
+between trust boundaries.
```

Comment:
is this supposed to be "across" the boundaries?

Reply:
I guess you could also phrase it that way -- the meaning is equivalent. I've usually seen "between" used here, for example "between kernel and userland", or "between host and guest memory".

Reply:
I mean "boundary" refers to the line between e.g. kernel and userland, so "between boundaries" would be unreasonable because you are crossing a line not transmitting between two lines.

Reply:
OK, I've tried rewording this sentence to be less ambiguous.